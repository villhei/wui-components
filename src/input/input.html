<style>
  input {
    --input-inactive-border-color: var(--grey-shade-4);
    --input-placeholder-color: var(--grey-shade-4);
    --input-focused-border-color: var(--primary-color);
  }

  input {
    transition: background-color ease-in-out var(--transition-base),
      border-color ease-in-out var(--transition-base);
  }

  label {
    transition: color ease-in-out var(--transition-base);
  }

  div {
    display: inline-flex;
    flex-direction: column;
    margin: var(--component-margin);
  }

  input {
    border-radius: var(--border-radius);
    font-size: var(--font-size);
    padding: 0.5rem 1rem;
    color: var(--text-color);
    border-color: var(--input-inactive-border-color);
    border-style: solid;
  }

  input:enabled:focus,
  input:enabled:focus-visible {
    outline-color: var(--input-focused-border-color);
    outline-offset: 0;
  }

  input::placeholder {
    color: var(--input-placeholder-color);
  }

  input:disabled {
    cursor: not-allowed;
    background-color: var(--grey-shade-5);
    color: var(--text-color-disabled);
  }

  input:enabled:invalid {
    border: var(--border-width) solid var(--warning-color);
  }

  input[variant="error"]:enabled {
    border: var(--border-width) solid var(--error-color);
  }

  input[variant="success"]:enabled {
    border: var(--border-width) solid var(--success-color);
  }

  label {
    margin-bottom: 0.5rem;
    text-transform: uppercase;
  }

  label:has(+ input:disabled) {
    color: var(--text-color-disabled);
  }

  label:has(+ input:enabled:focus) {
    color: var(--primary-color);
  }

  label:has(+ input[variant="error"]:enabled:not(:focus)) {
    color: var(--error-color);
  }

  label:has(+ input[variant="success"]:enabled:not(:focus)) {
    color: var(--success-color);
  }

  label:has(+ input:invalid) {
    color: var(--warning-color);
  }
</style>
<div>
  <label> </label>
  <input />
</div>
