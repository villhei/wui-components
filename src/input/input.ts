import templateContent from "bundle-text:./input.html";
import { WUIBase } from "../core";

const template = document.createElement("template");

template.innerHTML = templateContent;

const variants = ["success", "warning", "error"];
export class Input extends WUIBase {
  constructor() {
    super(template);
    this.inputNode = this.root.querySelector("input")!;
    this.labelNode = this.root.querySelector("label")!;
    this.messageNode = this.root.querySelector("#message")!;

    this.update();
  }

  inputNode: HTMLInputElement;
  labelNode: HTMLLabelElement;
  messageNode: HTMLElement;

  static get observedAttributes() {
    return ["class", "label", "name", "disabled", "value", "type", "message"];
  }

  get label() {
    return this.getAttribute("label");
  }

  get name() {
    return this.getAttribute("name");
  }

  get disabled() {
    return this.getBoolean("disabled");
  }

  get placeholder() {
    return this.getAttribute("placeholder");
  }

  get classes() {
    return this.getAttribute("class")?.split(",") || [];
  }

  get value() {
    return this.getAttribute("value") || "";
  }

  get type() {
    return this.getAttribute("type") || "text";
  }

  get variant() {
    return this.getAttribute("variant") || "default";
  }

  get message() {
    return this.getAttribute("message");
  }

  connectedCallback() {
    if (this.value) {
      this.inputNode.setAttribute("value", this.value);
    }
  }

  update() {
    this.classes.forEach((className) =>
      this.inputNode.classList.add(className)
    );
    this.labelNode.textContent = this.label;
    if (this.name) {
      this.labelNode.setAttribute("for", this.name);
      this.inputNode.setAttribute("name", this.name);
    }
    this.forwardAttribute("placeholder", this.inputNode);
    this.forwardAttribute("disabled", this.inputNode);
    this.forwardAttribute("type", this.inputNode);
    this.forwardAttribute("variant", this.inputNode);

    if (this.message) {
      this.messageNode.textContent = this.message;
    } else {
      this.messageNode.textContent = "";
    }
  }
}
