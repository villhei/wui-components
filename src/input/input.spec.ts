import { Input } from "./input";
import { fireEvent } from "@testing-library/dom";

const attachInternalsMock = jest.fn(() => {
  return {
    setFormValue: jest.fn(),
  };
});
beforeAll(() => {
  customElements.define("wui-input", Input);
});

beforeEach(() => {
  (Element.prototype as any).attachInternals = attachInternalsMock;
});

afterEach(() => {
  jest.resetAllMocks();
});

it("should match snapshot", () => {
  const el = new Input();
  expect(el.shadowRoot?.innerHTML).toMatchSnapshot();
});

it("should set label", () => {
  const el = new Input();
  el.setAttribute("label", "MyLabel");
  expect(el.shadowRoot?.querySelector("label")!).toHaveTextContent("MyLabel");
});

it("should set initial value", () => {
  const el = new Input();
  el.setAttribute("value", "SomeValue");
  el.connectedCallback();
  expect(el.shadowRoot?.querySelector("input")!).toHaveValue("SomeValue");
});

it("should allow listening for input events", () => {
  const el = new Input();
  el.connectedCallback();
  const listener = jest.fn();
  el.addEventListener("change", listener);
  const innerInput = el.shadowRoot?.querySelector("input")!;
  const event = { target: { value: "MyValue" } };
  fireEvent.change(innerInput, event);

  expect(listener).toHaveBeenCalledWith(
    expect.objectContaining({
      target: el,
    })
  );
});

it("should record internal input value as it's value", () => {
  const el = new Input();
  el.connectedCallback();
  const listener = jest.fn();
  el.addEventListener("change", listener);
  const innerInput = el.shadowRoot?.querySelector("input")!;
  const event = { target: { value: "MyValue" } };
  fireEvent.change(innerInput, event);
  expect(el.value).toEqual("MyValue");
});
