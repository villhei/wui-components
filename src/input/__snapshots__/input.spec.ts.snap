// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
"<style>
  input {
    --input-inactive-border-color: var(--grey-shade-4);
    --input-placeholder-color: var(--grey-shade-4);
    --input-focused-border-color: var(--primary-color);
  }

  input {
    transition: background-color ease-in-out var(--transition-base),
      border-color ease-in-out var(--transition-base);
  }

  label {
    transition: color ease-in-out var(--transition-base);
  }

  :host {
    display: inline-flex;
    flex-direction: column;
    padding: var(--component-padding);
    width: calc(100% - var(--component-padding) * 2);
  }

  input {
    border-radius: var(--border-radius);
    font-size: var(--font-size);
    padding: 0.5rem 1rem;
    color: var(--text-color);
    border-color: var(--input-inactive-border-color);
    border-style: solid;
  }

  input:enabled:focus,
  input:enabled:focus-visible {
    border-color: var(--input-focused-border-color);
    outline: none;
  }

  input::placeholder {
    color: var(--input-placeholder-color);
  }

  input:disabled {
    cursor: not-allowed;
    background-color: var(--grey-shade-5);
    color: var(--text-color-disabled);
  }

  input:enabled:read-only {
    font-style: italic;
    border-color: var(--input-inactive-border-color);
    background-color: var(--background-color);
    color: var(--text-color-disabled);
  }

  input:enabled:invalid {
    border: var(--border-width) solid var(--warning-color);
  }

  input[variant="error"]:enabled {
    border: var(--border-width) solid var(--error-color);
  }

  input[variant="success"]:enabled {
    border: var(--border-width) solid var(--success-color);
  }

  label {
    margin-bottom: 0.5rem;
    text-transform: uppercase;
  }

  label:has(+ input:enabled:read-only:focus) {
    color: var(--text-color);
  }

  label:has(+ input:disabled) {
    color: var(--text-color-disabled);
  }

  label:has(+ input:enabled:focus) {
    color: var(--primary-color);
  }

  label:has(+ input[variant="error"]:enabled:not(:focus)) {
    color: var(--error-color);
  }

  label:has(+ input[variant="success"]:enabled:not(:focus)) {
    color: var(--success-color);
  }

  label:has(+ input:invalid) {
    color: var(--warning-color);
  }

  #message {
    transition: max-height ease-in-out var(--transition-base);
  }

  #message {
    max-height: unset;
    font-size: 0.75rem;
    margin: 0.5rem 0 0.5rem 0.125rem;
  }

  #message:empty {
    max-height: 0;
    display: none;
  }

  input:enabled:invalid + #message {
    color: var(--warning-color);
  }

  input[variant="success"] + #message {
    color: var(--success-color);
  }

  input[variant="error"] + #message {
    color: var(--error-color);
  }
</style>
<label></label>
<input>
<p id="message"></p>
"
`;
